### PR TITLE
Structure splitting: Smoke test and refactor for numba

### LINKED ISSUE #158: `split_cells` expects `Sequence[Point]` but receives `ndarray`, causes error
When running the example in the README (or using `cellfinder-napari`), I get the following error:

```
  File "/home/adam/projects/brainglobe/cellfinder-core/src/cellfinder_core/main.py", line 69, in main
    points = detect.main(
  File "/home/adam/projects/brainglobe/cellfinder-core/src/cellfinder_core/detect/detect.py", line 167, in main
    cells = mp_3d_filter.process(async_results, callback=callback)
  File "/home/adam/projects/brainglobe/cellfinder-core/src/cellfinder_core/detect/filters/volume/volume_filter.py", line 102, in process
    return self.get_results()
  File "/home/adam/projects/brainglobe/cellfinder-core/src/cellfinder_core/detect/filters/volume/volume_filter.py", line 147, in get_results
    cell_centres = split_cells(
  File "/home/adam/projects/brainglobe/cellfinder-core/src/cellfinder_core/detect/filters/volume/structure_splitting.py", line 132, in split_cells
    xs = np.array([p.x for p in cell_points])  # TODO: use dataframe
  File "/home/adam/projects/brainglobe/cellfinder-core/src/cellfinder_core/detect/filters/volume/structure_splitting.py", line 132, in <listcomp>
    xs = np.array([p.x for p in cell_points])  # TODO: use dataframe
AttributeError: 'numpy.ndarray' object has no attribute 'x'
```

`cellfinder_core.detect.filters.volume.structure_splitting.split_cells` expects `Sequence[Point]` based on the type hints, but receives a numpy array, which doesn't have the correct attributes to be processed.

### LINKED ISSUE #159: More translation to `numba` needed in cell splitting code?
While reviewing #139, with a local fix to #158 (naive converting of the `np.array` to a list of `Point`, added the below to `split_cells`) 
```python
if not isinstance(cell_points[0],Point):
    cell_points = np.array([Point(p[0], p[1], p[0]) for p in cell_points])
```
I get a further error that I am finding hard-to-interpret (the joys of `numba`). This could be because my local fix to #158 was silly?
But also could be because we missed to translate some things to `numba`?

```python
Exception has occurred: TypingError
Failed in nopython mode pipeline (step: nopython frontend)
[1m[1m[1m[1m- Resolution failure for literal arguments:
[1mFailed in nopython mode pipeline (step: nopython frontend)
[1m[1m[1m[1m- Resolution failure for literal arguments:
[1mFailed in nopython mode pipeline (step: nopython frontend)
[1mUntyped global name 'get_structure_centre':[0m [1m[1mCannot determine Numba type of <class 'function'>[0m
[1m
File "src/cellfinder_core/detect/filters/volume/structure_detection.py", line 312:[0m
[1m    def structures_to_cells(self) -> List[Point]:
        <source elided>
            structure = iterator_pair.second
[1m            p = get_structure_centre(structure)
[0m            [1m^[0m[0m
[0m[0m
[0m[1m- Resolution failure for non-literal arguments:
[1mNone[0m
[0m[0m
[0m[1mDuring: resolving callee type: BoundFunction((<class 'numba.core.types.misc.ClassInstanceType'>, 'structures_to_cells') for instance.jitclass.CellDetector#7f918921d760<SOMA_CENTRE_VALUE:uint64,z:uint64,relative_z:uint64,next_structure_id:uint64,shape:UniTuple(int64 x 2),obsolete_ids:DictType[int64,int64]<iv=None>,coords_maps:DictType[uint64,array(uint64, 2d, A)]<iv=None>,previous_plane:array(uint64, 2d, A)>)[0m
[0m[1mDuring: typing of call at /home/alessandro/dev/cellfinder-core/src/cellfinder_core/detect/filters/volume/structure_detection.py (233)
[0m
[1m
File "src/cellfinder_core/detect/filters/volume/structure_detection.py", line 233:[0m
[1m    def get_cell_centres(self) -> List[Point]:
[1m        cell_centres = self.structures_to_cells()
[0m        [1m^[0m[0m
[0m
[0m[1m- Resolution failure for non-literal arguments:
[1mNone[0m
[0m[0m
[0m[1mDuring: resolving callee type: BoundFunction((<class 'numba.core.types.misc.ClassInstanceType'>, 'get_cell_centres') for instance.jitclass.CellDetector#7f918921d760<SOMA_CENTRE_VALUE:uint64,z:uint64,relative_z:uint64,next_structure_id:uint64,shape:UniTuple(int64 x 2),obsolete_ids:DictType[int64,int64]<iv=None>,coords_maps:DictType[uint64,array(uint64, 2d, A)]<iv=None>,previous_plane:array(uint64, 2d, A)>)[0m
[0m[1mDuring: typing of call at <string> (3)
[0m
[1m
File "<string>", line 3:[0m
[1m<source missing, REPL/exec in use?>[0m
  File "/home/alessandro/dev/cellfinder-core/src/cellfinder_core/detect/filters/volume/structure_splitting.py", line 82, in ball_filter_imgs
    return ball_filtered_volume, cell_detector.get_cell_centres()
  File "/home/alessandro/dev/cellfinder-core/src/cellfinder_core/detect/filters/volume/structure_splitting.py", line 97, in iterative_ball_filter
    vol, cell_centres = ball_filter_imgs(
  File "/home/alessandro/dev/cellfinder-core/src/cellfinder_core/detect/filters/volume/structure_splitting.py", line 157, in split_cells
    ns, centres = iterative_ball_filter(vol)
  File "/home/alessandro/dev/cellfinder-core/src/cellfinder_core/detect/filters/volume/volume_filter.py", line 151, in get_results
    cell_centres = split_cells(
  File "/home/alessandro/dev/cellfinder-core/src/cellfinder_core/detect/filters/volume/volume_filter.py", line 109, in process
```

### LINKED ISSUE #166: Test for cell splitting code
We should have a test that covers the issue described in #158

### PR DESCRIPTION
**Pull request by alessandrofelder**
## Description

This PR adds 
- an integration smoke test and additional test data (created from a real-life dataset) to reproduce #158 and cover our structure splitting code with the tests.
- a refactor to consistently use `np.ndarray`s and other numba-suitable refactoring in the structure splitting code.

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**

Our tests don't currently cover the case where cell detection requires structure splitting.

**What does this PR do?**

See description.

## References

Closes #166, closes #158, closes #159 

## How has this PR been tested?
Run tests locally and on CI.

## Does this PR require an update to the documentation?

New test and fixtures have docstrings, test data documented in docstring of new testing module.

## Checklist:

- [x] The code has been tested locally
- [ ] Tests have been added to cover all new functionality (unit & integration)
- [ ] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

**Comment by alessandrofelder** [comment]
Note that CI fails in expected way, ensuring our tests now can reproduce #158.

**Comment by adamltyson** [comment]
@alessandrofelder would it be worth renaming the directory containing the new test data? The dataset label doesn't mean much to anyone other than me. Maybe something like `structure_split_test`, then just add a note (maybe a txt file) to highlight which dataset it's from for our records?

File: src/cellfinder_core/detect/filters/volume/structure_detection.py [hunk 1]
```diff
@@ -45,34 +45,20 @@ def traverse_dict(d: Dict[T, T], a: T) -> T:
         return a
 
 
-def get_structure_centre(structure: Sequence[Point]) -> Point:
+@njit
+def get_structure_centre(structure: np.ndarray) -> np.ndarray:
```
Comment (dstansby) [inline src/cellfinder_core/detect/filters/volume/structure_detection.py:49]: It looks like this was written with cython in mind, and can now be re-written much more simply as just 
```python
return np.round(np.mean(structure, axis=0))
```
?

File: src/cellfinder_core/detect/filters/volume/structure_detection.py [hunk 2]
```diff
@@ -272,12 +257,12 @@ def merge_structures(
...
-        cell_centres = []
-        for iterator_pair in self.coords_maps:
-            structure = iterator_pair.second
+    def structures_to_cells(self) -> np.ndarray:
+        cell_centres = np.empty((len(self.coords_maps.keys()),3))
+        idx = 0
+        for structure_id, structure in self.coords_maps.items():
             p = get_structure_centre(structure)
-            cell_centres.append(p)
+            cell_centres[idx] = p
```
Comment (dstansby) [inline src/cellfinder_core/detect/filters/volume/structure_detection.py:265]: Is this missing an increment of `idx`? It looks like the only index written to (and repeatedly overitten) at the moment is 0.

File: src/cellfinder_core/detect/filters/volume/structure_splitting.py [hunk 3]
```diff
@@ -111,14 +111,14 @@ def iterative_ball_filter(
...
-def check_centre_in_cuboid(centre: Point, max_coords: np.ndarray) -> bool:
+def check_centre_in_cuboid(centre: np.ndarray, max_coords: np.ndarray) -> bool:
     """
     Checks whether a coordinate is in a cuboid
     :param centre: x,y,z coordinate
     :param max_coords: far corner of cuboid
     :return: True if within cuboid, otherwise False
     """
-    relative_coords = np.array([centre.x, centre.y, centre.z])
+    relative_coords = np.array([centre[0], centre[1], centre[2]])
```
Comment (dstansby) [inline src/cellfinder_core/detect/filters/volume/structure_splitting.py:121]:
```suggestion
    relative_coords = centre
```
Casting to an array is now redundant as it's already an array

File: src/cellfinder_core/detect/filters/volume/structure_splitting.py [hunk 4]
```diff
@@ -130,23 +130,24 @@ def check_centre_in_cuboid(centre: Point, max_coords: np.ndarray) -> bool:
...
+    cell_points: np.ndarray, outlier_keep: bool = False
 ) -> List[Point]:
-    orig_centre = get_structure_centre_wrapper(cell_points)
+    
+    orig_centre = get_structure_centre(cell_points)
 
-    xs = np.array([p.x for p in cell_points])  # TODO: use dataframe
-    ys = np.array([p.y for p in cell_points])
-    zs = np.array([p.z for p in cell_points])
+    xs = np.array([p[0] for p in cell_points])  # TODO: use dataframe
```
Comment (dstansby) [inline src/cellfinder_core/detect/filters/volume/structure_splitting.py:138]:
```suggestion
    xs = p[:, 0]
```
etc. for y and z

File: src/cellfinder_core/detect/filters/volume/volume_filter.py [hunk 5]
```diff
@@ -146,10 +147,11 @@ def get_results(self) -> List[Cell]:
             cell_volume = len(cell_points)
 
             if cell_volume < max_cell_volume:
-                cell_centre = get_structure_centre_wrapper(cell_points)
+                cell_centre = get_structure_centre(cell_points)
+                cell_centre_point = Point(cell_centre[0], cell_centre[1], cell_centre[2])
```
Comment (dstansby) [inline src/cellfinder_core/detect/filters/volume/volume_filter.py:151]: Not sure why a point is being constructed here, but not in the lines changed below? I'd say don't bother using `Point` here.

File: tests/tests/test_unit/test_detect/test_filters/test_volume_filters/test_structure_detection.py [hunk 6]
```diff
@@ -50,10 +50,17 @@ def structure(three_d_cross):
     s = [Point(*c) for c in zip(coords[0], coords[1], coords[2])]
     return s
 
+@pytest.fixture()
+def structure_as_array(structure):
```
Comment (dstansby) [inline tests/tests/test_unit/test_detect/test_filters/test_volume_filters/test_structure_detection.py:54]: Would be good to add some typing to this

**Review by dstansby — CHANGES_REQUESTED** [review]
Just to double check, everyone is happy putting another 20MB of sample data in this repository? It will live here forever (unless someone re-writes the git history), so maybe now is the time to sort out sample data downloading with `pooch` instead of adding this new data to the repository?

**Comment by alessandrofelder** [comment]
Thanks for your comments, both!

Re test data, I had that thought too, but decided against it as 
- the new test data is on the same scale as existing test data
- it would be nice to get a new release over the line
But maybe it's worth reconsidering this decision?
We definitely want to be using `pooch` soon!

**Comment by dstansby** [comment]
👍 sounds good

File: src/cellfinder_core/detect/filters/volume/structure_detection.py (hunk 2 above)
Comment (alessandrofelder) [reply src/cellfinder_core/detect/filters/volume/structure_detection.py:265]: 😱 good catch!

File: src/cellfinder_core/detect/filters/volume/structure_splitting.py (hunk 4 above)
Comment (alessandrofelder) [reply src/cellfinder_core/detect/filters/volume/structure_splitting.py:138]: (actually `cell_points[:, 0]`)

File: src/cellfinder_core/detect/filters/volume/structure_detection.py (hunk 1 above)
Comment (alessandrofelder) [reply src/cellfinder_core/detect/filters/volume/structure_detection.py:49]: [Numba doesn't support `axis`](https://numba.pydata.org/numba-doc/dev/reference/numpysupported.html#calculation) but I've simplified along these lines.

File: src/cellfinder_core/detect/filters/volume/structure_splitting.py [hunk 7]
```diff
@@ -167,20 +171,24 @@ def split_cells(
...
+    for i, relative_centre in enumerate(relative_centres):
+        absolute_centre = np.array(
+            [
+                orig_corner[0] + relative_centre[0],
+                orig_corner[1] + relative_centre[1],
+                orig_corner[2] + relative_centre[2],
+            ]
         )
-        absolute_centres.append(absolute_centre)
+        absolute_centres[i] = absolute_centre
```
Comment (dstansby) [inline src/cellfinder_core/detect/filters/volume/structure_splitting.py:192]: Might be a bit nicer to write this as the following, which is a bit shorter and avoids allocating a new array.
```python
absolute_centres[:, 0] = orig_corner[0] + relative_centre[0]
absolute_centres[:, 1] = orig_corner[1] + relative_centre[1]
absolute_centres[:, 2] = orig_corner[2] + relative_centre[2]
```

File: src/cellfinder_core/detect/filters/volume/structure_splitting.py [hunk 8]
```diff
@@ -157,7 +161,7 @@ def split_cells(
     # centres is a list of lists of centres (1 list of centres per ball run)
     ns, centres = iterative_ball_filter(vol)
     ns.insert(0, 1)
-    centres.insert(0, [relative_orig_centre])
+    centres.insert(0, relative_orig_centre)
```
Comment (dstansby) [inline src/cellfinder_core/detect/filters/volume/structure_splitting.py:164]: This is now not a "list of lists of centres" as the comment above on line 161 says it should be - given the return of `iterative_ball_filter` hasn't changed, I think this *should* stay as `[relative_orig_centre]`?

I am guessing you changed this for a reason though... any thoughts?

**Review by dstansby — COMMENTED** [review]
I've left one major comment that I think needs resolving about `relative_orig_centre`, but otherwise this looks 👍

File: src/cellfinder_core/detect/filters/volume/structure_splitting.py (hunk 8 above)
Comment (alessandrofelder) [reply src/cellfinder_core/detect/filters/volume/structure_splitting.py]: AFAICT the return of `iterative_ball_filter` did change from `List[List[Point]]` to `[List[np.ndarray]`, so `mypy` was complaining about the typing with the previous `[relative_orig_centre]`, which was the reason I removed it.

However, I hadn't thought things through, and therefore this is still a good catch of a bug I introduced! I've adapted the comment and fixed the bug to reflect my intentions now:
```python
# centres is a list of arrays of centres (1 array of centres per ball run)
ns, centres = iterative_ball_filter(vol)
ns.insert(0, 1)
centres.insert(0, np.array([relative_orig_centre]))
```

**Review by dstansby — APPROVED** [review]

**Comment by dstansby** [comment]
Nice - hopefully code coverage should increase when this commit runs on the `main` branch!

**Pull request merged**